feat: expose package_annotations

betaboon opened the pull request:
I'm unsure about naming and the location of `pycross_package_annotation`, so i would be happy about some feedback there.

The approach of handing over json-files is inspired by [rules_python's package_annotation](https://github.com/bazelbuild/rules_python/blob/main/python/pip_install/pip_repository.bzl#L1033)

for some reason i can't get `bazel build //docs/...` working on my nixos-system. so i guess that's why CI fails here.

betaboon commented on pycross/private/package_annotation.bzl:
```diff
@@ -0,0 +1,24 @@
+def pycross_package_annotation(
+        always_build = False,
+        build_dependencies = [],
+        build_target_override = None,
```
maybe this should be renamed to `build_target` to be symmetric to `lock_import.package` ?
if so, maybe that change should be reflected in `class PackageAnnotation`

betaboon replied:
done

jvolkman reviewed (commented):
This is nice. Thanks!

I think you could place `pycross_package_annotation` in lock_attrs.bzl since it's really just a rich attribute value. And maybe drop the `pycross` from that one and just call it `package_annotation`.

betaboon commented:
i renamed and moved stuff according to the comments.
i included the `install_exclude_globs` PR here.

still can't get the docs-generation working locally. could you do that?

jvolkman commented on pycross/private/tools/wheel_installer.py:
```diff
@@ -39,6 +45,31 @@ def setup_namespace_pkg_compatibility(wheel_dir: Path) -> None:
+
+    @classmethod
+    @contextmanager
+    def open(cls, path: "os.PathLike[str]", install_exclude_globs: List[str]) -> Iterator["FilteredWheelFile"]:
```
add
```
from __future__ import annotations

from typing import Union
```

and this can be
```
def open(cls, path: Union[os.PathLike, str], install_exclude_globs: List[str]) -> Iterator[FilteredWheelFile]:
```

betaboon replied:
done

jvolkman reviewed (commented):
This is great, thanks! One minor comment. Also, would you mind adding some entries to CHANGELOG.md? An 'added' entry for `install_exclude_globs`, and a breaking change entry for the things that `package_annotations` replaces.
